src/prebid.ts

```diff
@@ -752,13 +751,18 @@ export const requestBids = (function() {
     // if the request does not specify adUnits, clone the global adUnit array;
     // otherwise, if the caller goes on to use addAdUnits/removeAdUnits, any asynchronous logic
     // in any hook might see their effects.
-    const req = options as PrivRequestBidsOptions;
-    const adUnits = req.adUnits || pbjsInstance.adUnits;
-    req.adUnits = (Array.isArray(adUnits) ? adUnits.slice() : [adUnits]);
+    const adUnits = options.adUnits || pbjsInstance.adUnits;
```

Comment:
I think the filtering of the adunits based on `adUnitCodes` should happen before the event.

Reply:
Hm, how would it be breaking? Isn't it only the metrics and this event which is being called in between ad units definition and the `adUnitCodes` filter?

Reply:
Ah ok. It would be great to do it like mentioned then.
> unless we do it here just for the event payload and then later again where we do it now, which is an option


If the event were used to add an ad unit, it would also need to add the ad unit code to `adUnitCodes` (if present).

Reply:
> If the event were used to add an ad unit, it would also need to add the ad unit code to adUnitCodes (if present).

I missed this remark - I am  not sure this makes sense; if I am selecting a specific ad unit with `adUnitCodes`, wouldn't it be surprising if some other code appears in the auction?

On the other hand it's also difficult to let event handlers modify `adUnitCodes` as they could modify it in a way that suggests they were supposed to see ad units that were filtered out before then.

After implementing a version of this, I am not so sure that it's better. It does make it less likely that an event handlers works on ad units it doesn't realize are going to be filtered out, but at the cost of no longer being able to do whatever it pleases with the request.